src/NSubstitute/Core/IReturn.cs

```diff
@@ -45,14 +46,14 @@ public ReturnValueFromFunc(Func<CallInfo, T> funcToReturnValue)
             _funcToReturnValue = funcToReturnValue ?? ReturnNull();
         }
 
-        public object ReturnFor(CallInfo info) => _funcToReturnValue(info);
+        public object? ReturnFor(CallInfo info) => _funcToReturnValue(info);
         public Type TypeOrNull() => typeof (T);
         public bool CanBeAssignedTo(Type t) => typeof (T).IsAssignableFrom(t);
 
         private static Func<CallInfo, T> ReturnNull()
         {
             if (typeof(T).GetTypeInfo().IsValueType) throw new CannotReturnNullForValueType(typeof(T));
-            return x => default(T);
+            return x => default(T)!;
```

Comment:
**question:** won't this always be null for ref types? I was expecting this to return `Func<CallInfo, T?>`. (sorry I haven't played with non-nullable ref types yet, so these may be silly questions)

Reply:
THANK YOU very much for this question! Apparently they made a change in .NET 5 regarding that.

Previously the following example would not work:

```c#
public class Data<T>
{
    public Data(T? value) { }
}
```

Compiler would have complained that it's unclear what is `T?`. For value types that should be `Nullable<T>`, while for reference type it would be just `T` with annotated nullability.

Apparently they changed it now, so if you don't specify that T is a value type, they assume you mean nullability for reference types. For value types is means nothing, as value type cannot be `null`.

OTOH is you specify `T` constrain, it would be using `Nullable<T>`:

```c#
public class Data<T> where T : struct
{
    public Data(T? value)
    { 
        bool isSet = value.HasValue;
    }
}
```

A bit obscure design, but I think they found that it would cater for a big variety of scenarios and harm is lesser than benefit.

Thank you for forcing me to re-check this! I'm happy I learned something new.

I'll revisit all my changes to make them aware of the change. By the time you are reading this most likely I've already pushed my changes 😅